### PR TITLE
[Merged by Bors] - feat: Port/Data.List.Intervals

### PR DESCRIPTION
**Pull request by zaxioms0**
feat: Port/data.list.intervals

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/List/Intervals.lean [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+  apply eq_nil_iff_forall_not_mem.2
+  intro a
+  simp only [and_imp, not_and, not_lt, List.mem_inter, List.Ico.mem]
+  intro _ h₂ h₃
+  exfalso
+  exact not_lt_of_ge h₃ h₂
+#align list.Ico.inter_consecutive List.Ico.inter_consecutive
+
+@[simp]
+theorem bagInter_consecutive (n m l : Nat) : @List.bagInter ℕ instBEq (Ico n m) (Ico m l) = [] :=
```
Comment (xoers) [inline Mathlib/Data/List/Intervals.lean:121]: can someone just double check this theorem still says what the original mathlib intended it to say? I needed to help it along with the implicits.

File: Mathlib/Data/List/Intervals.lean [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+
+This file defines intervals of naturals. `list.Ico m n` is the list of integers greater than `m`
+and strictly less than `n`.
+
+## TODO
+- Define `Ioo` and `Icc`, state basic lemmas about them.
+- Also do the versions for integers?
+- One could generalise even further, defining 'locally finite partial orders', for which
+  `Set.Ico a b` is `[finite]`, and 'locally finite total orders', for which there is a list model.
+- Once the above is done, get rid of `Data.Int.range` (and maybe `list.range'`?).
```
Comment (jcommelin) [inline Mathlib/Data/List/Intervals.lean:25]:
```suggestion
- Once the above is done, get rid of `Data.Int.range` (and maybe `List.range'`?).
```

File: Mathlib/Data/List/Intervals.lean [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+
+open Nat
+
+namespace List
+
+/-- `Ico n m` is the list of natural numbers `n ≤ x < m`.
+(Ico stands for "interval, closed-open".)
+
+See also `data/set/intervals.lean` for `set.Ico`, modelling intervals in general preorders, and
+`multiset.Ico` and `finset.Ico` for `n ≤ x < m` as a multiset or as a finset.
```
Comment (jcommelin) [inline Mathlib/Data/List/Intervals.lean:37]:
```suggestion
See also `Data/Set/Intervals.lean` for `Set.Ico`, modelling intervals in general preorders, and
`Multiset.Ico` and `Finset.Ico` for `n ≤ x < m` as a multiset or as a finset.
```

File: Mathlib/Data/List/Intervals.lean [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+#align list.Ico List.Ico
+
+namespace Ico
+
+theorem zero_bot (n : ℕ) : Ico 0 n = range n := by rw [Ico, tsub_zero, range_eq_range']
+#align list.Ico.zero_bot List.Ico.zero_bot
+
+@[simp]
+theorem length (n m : ℕ) : length (Ico n m) = m - n :=
+  by
```
Comment (jcommelin) [inline Mathlib/Data/List/Intervals.lean:50]: Please move `by` to the end of the statement. Also elsewhere.
```suggestion
theorem length (n m : ℕ) : length (Ico n m) = m - n := by
  by
```

Comment (xoers) [reply Mathlib/Data/List/Intervals.lean]: should be fixed now!

**Comment by xoers** [comment]
Makes sense to me. I'll take a look at fixing it in the morning

**Comment by nomeata** [comment]
@zaxioms this is labeled WIP; are you actively working on it? Maybe use labels help-wanted or please-adopt if that’s closer to the truth

**Comment by zaxioms** [comment]
> @zaxioms this is labeled WIP; are you actively working on it? Maybe use labels help-wanted or please-adopt if that’s closer to the truth

It's done except for the issue resolving `instBEqNat` vs. `instBEq`. I'm not entirely sure how to resolve this issue, but from my understanding [this PR](https://github.com/leanprover/lean4/pull/2038) will fix the issue as mentioned [here](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/instBEqNat.20.3D.20instBEq.20.3A.3D.20rfl.20--fails/near/321449400). I didn't have the help wanted tag because I'm not sure there's much to be done until this is resolved, but I could be wrong and if you think I should I can add it

**Comment by nomeata** [comment]
(I'm not an expert, so I hope others can advise)

**Comment by zaxioms** [comment]
ahh I was mistaken, I checked randomly, but I forgot I made changes other places locally that made it work for this one case

**Comment by zaxioms** [comment]
added tag because it seems the PR hasn't gone through yet so I'm wondering if there's another work around

**Pull request closed without merging**